### ISSUE TITLE
Training results in bird dataset is not good as flowers.

### ISSUE THREAD

**Issue by Xuchenfeng714** [aelnouby/Text-to-Image-Synthesis#8]
Hi!When I trained the bird dataset,I found the results is not good.Have you trained the bird dataset and how about the result? or do you have any advice to train the birds?
Hope for reply!

**Comment by aelnouby** [comment]
I have trained the birds dataset, and indeed the results are not as good as the flowers, they are not terrible but most of the images are not very convincing.

I would suggest trying WGAN with gradient penalty, at the time of development of this project, Pytorch had a bug with second order gradients so I could not run Gradient penalty, but it should be fixed now. 

Install Pytorch from source (latest master branch), in the trainer.py file you will find commented code calling gradient penalty, uncomment it and comment weight clipping lines.

Hope this helps, good luck

**Comment by Xuchenfeng714** [comment]
> Thanks a lot!

**Comment by Xuchenfeng714** [comment]
I have tried WGAN with gradient penalty,but the loss rise rapidly.
Epoch: 0, Gen_iteration: 5, d_loss= 103.440140, g_loss= 56.608669, real_loss= 52.557484, fake_loss = -50.882660
Epoch: 1, Gen_iteration: 10, d_loss= 863.011414, g_loss= 449.037720, real_loss= 481.742889, fake_loss = -381.268524
Epoch: 2, Gen_iteration: 15, d_loss= 2133.874756, g_loss= 1025.139893, real_loss= 1100.198242, fake_loss = -1033.676514
Epoch: 3, Gen_iteration: 20, d_loss= 3898.405518, g_loss= 1930.657104, real_loss= 1996.155396, fake_loss = -1902.250122
Epoch: 4, Gen_iteration: 25, d_loss= 5739.763184, g_loss= 2658.287109, real_loss= 2969.346924, fake_loss = -2770.416260
Epoch: 5, Gen_iteration: 111, d_loss= 8150.391602, g_loss= 3850.994385, real_loss= 4235.759277, fake_loss = -3914.632080
Epoch: 6, Gen_iteration: 197, d_loss= 10824.121094, g_loss= 5187.048340, real_loss= 5469.530762, fake_loss = -5354.590332
Epoch: 7, Gen_iteration: 283, d_loss= 13580.561523, g_loss= 6707.688477, real_loss= 7218.591797, fake_loss = -6361.969727
Epoch: 8, Gen_iteration: 369, d_loss= 16937.488281, g_loss= 7948.286621, real_loss= 8532.394531, fake_loss = -8405.094727
Epoch: 9, Gen_iteration: 455, d_loss= 20207.320312, g_loss= 9759.650391, real_loss= 10420.387695, fake_loss = -9786.931641
Epoch: 10, Gen_iteration: 522, d_loss= 24503.666016, g_loss= 11703.536133, real_loss= 12888.699219, fake_loss = -11614.966797
Epoch: 11, Gen_iteration: 608, d_loss= 28469.443359, g_loss= 13522.029297, real_loss= 14489.694336, fake_loss = -13979.749023
Epoch: 12, Gen_iteration: 694, d_loss= 31841.078125, g_loss= 15438.813477, real_loss= 16013.761719, fake_loss = -15827.315430
Epoch: 13, Gen_iteration: 780, d_loss= 36917.523438, g_loss= 17693.416016, real_loss= 18982.197266, fake_loss = -17935.326172
Epoch: 14, Gen_iteration: 866, d_loss= 41574.867188, g_loss= 20319.597656, real_loss= 21622.007812, fake_loss = -19952.857422
Epoch: 15, Gen_iteration: 952, d_loss= 46522.523438, g_loss= 23101.357422, real_loss= 24239.470703, fake_loss = -22283.054688
Epoch: 16, Gen_iteration: 1019, d_loss= 51993.117188, g_loss= 25345.166016, real_loss= 26735.267578, fake_loss = -25257.851562
Epoch: 17, Gen_iteration: 1105, d_loss= 55917.226562, g_loss= 27994.000000, real_loss= 28925.216797, fake_loss = -26992.009766
Epoch: 18, Gen_iteration: 1191, d_loss= 62476.812500, g_loss= 30643.603516, real_loss= 32183.089844, fake_loss = -30293.722656
Epoch: 19, Gen_iteration: 1277, d_loss= 67617.656250, g_loss= 33088.289062, real_loss= 34308.250000, fake_loss = -33309.402344
Epoch: 20, Gen_iteration: 1363, d_loss= 74095.156250, g_loss= 35776.804688, real_loss= 37719.585938, fake_loss = -36375.566406
Epoch: 21, Gen_iteration: 1449, d_loss= 71557.281250, g_loss= 33453.023438, real_loss= 36513.445312, fake_loss = -35043.832031
Epoch: 22, Gen_iteration: 1516, d_loss= 85630.593750, g_loss= 42493.531250, real_loss= 43901.222656, fake_loss = -41729.367188
Epoch: 23, Gen_iteration: 1602, d_loss= 94216.078125, g_loss= 46028.574219, real_loss= 47923.339844, fake_loss = -46292.734375
Epoch: 24, Gen_iteration: 1688, d_loss= 100888.734375, g_loss= 48869.953125, real_loss= 50821.664062, fake_loss = -50067.074219
Epoch: 25, Gen_iteration: 1774, d_loss= 107203.671875, g_loss= 52571.347656, real_loss= 54481.753906, fake_loss = -52721.914062
Epoch: 26, Gen_iteration: 1860, d_loss= 115674.226562, g_loss= 55484.093750, real_loss= 58041.207031, fake_loss = -57633.019531

hope for reply for the way to improve it.

**Comment by aelnouby** [comment]
Make sure you have turned off the weight clipping, also use a lower learning rate 0.00005 for example

**Comment by Xuchenfeng714** [comment]
Yes. In the trained.py.I followed your advice 
 fake_images = Variable(self.generator(right_embed, noise).data)
                    outputs, _ = self.discriminator(fake_images, right_embed)
                    fake_loss = torch.mean(outputs)
                    fake_loss.backward(one)

                    ## NOTE: Pytorch had a bug with gradient penalty at the time of this project development
                    ##       , uncomment the next two lines and remove the params clamping below if you want to try gradient penalty
                    gp = Utils.compute_GP(self.discriminator, right_images.data, right_embed, fake_images.data, LAMBDA=10)
                    gp.backward()

                    d_loss = real_loss - fake_loss + gp

                    if cls:
                        d_loss = d_loss - wrong_loss

                    self.optimD.step()

                   # for p in self.discriminator.parameters():
                   #    p.data.clamp_(-0.01, 0.01)

And changed the "improve=False"to"improve =True" in the wgan.py. then the rapidly loss rising occur.

**Comment by aelnouby** [comment]
Did you change the learning rate to 0.00005 as well ?

**Comment by Xuchenfeng714** [comment]
Yes,I changed the learning rate to 0.00005 and the result is better now but it is still not as good as the result birds of the author version

**Comment by Xuchenfeng714** [comment]
The author version is through LUA while we use PYTHON and we have the same model "DCGAN“，but the bird result is still not as good as the author version even I have tried many ways to improve it such as data augmentation,changing the optimization,changing the size of the real image from 64*64 to 128*128,and changing the learning rate and so on.Will the different programming language lead to the different result?I am so confused.

**Comment by aelnouby** [comment]
I don't think the programming language should change anything, however be aware that usually in papers authors cherry-pick the best examples they could get, this doesn't mean that all the samples were of the same quality, this is particularly a big problem with GANs since there is not any objective way of evaluating images quality, so you cannot really know how a method is doing. 

If you really want to be sure you might want to try the authors Lua code, only then you can know if the images quality is as good as reported or not.

**Comment by Xuchenfeng714** [comment]
OK!Thanks for your reply! I learn a lot from you!

**Comment by aelnouby** [comment]
You are very welcome, thanks for your comments, Good luck
